internal/controller/nonadmindownloadrequest_controller.go

```diff
@@ -0,0 +1,188 @@
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.20.2/pkg/reconcile
+//
+// This reconcile implements ObjectReconciler https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.20.2/pkg/reconcile#ObjectReconciler
+// Each reconciliation event gets the associated object from Kubernetes before passing it to Reconcile
+func (r *NonAdminDownloadRequestReconciler) Reconcile(ctx context.Context, req *nacv1alpha1.NonAdminDownloadRequest) (reconcile.Result, error) {
+	if req == nil || req.Spec.Target.Kind == constant.EmptyString {
+		return ctrl.Result{Requeue: false}, nil
+	}
+	// TODO: delete if UID is always available here.
+	// if req.ObjectMeta.UID  == "" {
+	// requeue later until uid is populated
+	// }
+	logger := log.FromContext(ctx)
+	logger.Info("Reconciling NonAdminDownloadRequest", "name", req.Name, "namespace", req.Namespace)
+	if req.Status.VeleroDownloadRequestStatus.Expiration.After(time.Now()) {
+		// request is expired, so delete NADR after deleting velero DR
+		// find associated downloadrequest and delete that first
+		logger.V(1).Info("Deleting expired NonAdminDownloadRequest", req.Name, req.Namespace)
+		if err := r.Delete(ctx, req); err != nil {
+			if apierrors.IsNotFound(err) {
+				return reconcile.Result{Requeue: false}, nil
+			}
+			return reconcile.Result{Requeue: true}, nil
+		}
+	}
+	var veleroObj, naObj client.Object
+	switch req.Spec.Target.Kind {
+	case velerov1.DownloadTargetKindBackupLog,
+		velerov1.DownloadTargetKindBackupContents,
+		velerov1.DownloadTargetKindBackupVolumeSnapshots,
+		velerov1.DownloadTargetKindBackupItemOperations,
+		velerov1.DownloadTargetKindBackupResourceList,
+		velerov1.DownloadTargetKindBackupResults,
+		velerov1.DownloadTargetKindCSIBackupVolumeSnapshots,
+		velerov1.DownloadTargetKindCSIBackupVolumeSnapshotContents,
+		velerov1.DownloadTargetKindBackupVolumeInfos:
+		// get velero backup name from nab
+		var nab nacv1alpha1.NonAdminBackup
+		if err := r.Get(ctx, types.NamespacedName{Namespace: r.OADPNamespace, Name: req.Spec.Target.Name}, &nab); err != nil {
+			// TODO:
+			_ = ""
+		}
+		naObj = &nab
+		if nab.VeleroBackupName() == constant.EmptyString {
+			return reconcile.Result{}, errors.New("unable to get Velero Backup UUID from NonAdminBackup Status")
+		}
+		var vb = velerov1.Backup{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      nab.VeleroBackupName(),
+				Namespace: r.OADPNamespace,
+			},
+		}
+		veleroObj = &vb
+	case velerov1.DownloadTargetKindRestoreLog,
+		velerov1.DownloadTargetKindRestoreResults,
+		velerov1.DownloadTargetKindRestoreResourceList,
+		velerov1.DownloadTargetKindRestoreItemOperations,
+		velerov1.DownloadTargetKindRestoreVolumeInfo:
+		veleroObj = &velerov1.Restore{}
+	}
+	// TODO:
+	_ = naObj
+	if err := r.Get(ctx, types.NamespacedName{Namespace: veleroObj.GetNamespace(), Name: veleroObj.GetName()}, veleroObj); err != nil {
+		if apierrors.IsNotFound(err) {
+			// create
+			if createErr := r.Create(ctx, veleroObj); createErr != nil {
```

Comment:
Why are we attempting to create the `veleroObj` here? If the backup doesn't exist, then shouldn't we just error out here?

Reply:
probably thinking about creating download requests.